File: src/main/java/tn/esprit/piproject/Entities/Company.java
package tn.esprit.piproject.Entities;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import com.fasterxml.jackson.databind.deser.std.StdDeserializer;
import com.mongodb.lang.Nullable;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import lombok.*;
import lombok.extern.jackson.Jacksonized;
import org.bson.types.Binary;
import org.springframework.data.mongodb.core.mapping.DBRef;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.Field;

import java.util.ArrayList;
import java.util.List;


@Document(collection = "companies")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@ToString
public class Company {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private int id;
    private String name;
    private String email;
    private String description;
    private String address;
    private int Pnumber;
    private String attachmentFileName;
    @JsonDeserialize(using = BinaryDeserializer.class)
    private Binary attachmentData;
    @DBRef
    private List<Offer> offers = new ArrayList<>();
}

File: PI24-mainnnnn/AngularPIFrontnnnnn/src/app/core/Company.ts
import { Binary } from "@angular/compiler";
import { Offer } from "./Offer";

export class Company {
    id: number;
    name: string;
    address: string;
    email: string;
    description: string;
    pnumber: Number;
    offers: Offer[];
    attachmentFileName?: string;
    attachmentData?: Binary;
};
function filterCompanies(companies: Company[], searchTerm: string): Company[] {
    if (!searchTerm) {
      return companies; // Return all companies if search term is empty
    }

    const searchTermLower = searchTerm.toLowerCase(); // Case-insensitive search

    return companies.filter(company => {
      return (
        company.name?.toLowerCase().includes(searchTermLower) ||
        company.address?.toLowerCase().includes(searchTermLower) ||
        company.email?.toLowerCase().includes(searchTermLower) ||
        company.description?.toLowerCase().includes(searchTermLower) ||
        company.pnumber?.toString().includes(searchTermLower)
      );
    });
  }
